src/lib.rs

```diff
@@ -184,57 +195,80 @@ impl SteamID {
-                .as_str()
-                .parse()
-                .unwrap(),
-        )
-        .ok_or_else(Self::err)?;
-        let account_id = groups
-            .get(3)
-            .ok_or_else(Self::err)?
-            .as_str()
-            .parse()
-            .unwrap();
-
-        let mut instance: Option<Instance> = groups
-            .get(5)
-            .map(|g| Instance::from_u64(g.as_str().parse().unwrap()).unwrap_or(Instance::Invalid));
-
-        if instance.is_none() && type_char == 'U' {
-            instance = Some(Instance::Desktop);
-        } else if type_char == 'T' || type_char == 'g' || instance.is_none() {
-            instance = Some(Instance::All);
+        if type_char != 'i' && type_char != 'I' && account_type == AccountType::Invalid {
+            return None;
+        }
+
+        if bytes.next() != Some(b':') {
+            return None;
+        }
+
+        let universe = bytes.next().and_then(digit_from_ascii).and_then(|digit| {
+            if digit <= 4 {
+                Universe::from_u64(u64::from(digit))
+            } else {
+                None
+            }
+        })?;
+
+        if bytes.next() != Some(b':') {
+            return None;
+        }
+
+        let mut account_id = u32::from(digit_from_ascii(bytes.next()?)?);
+        while let Some(digit) = bytes.peek().copied().and_then(digit_from_ascii) {
+            bytes.next().expect("Byte was peeked");
+            account_id = account_id.checked_mul(10)?;
+            account_id = account_id.checked_add(u32::from(digit))?;
         }
 
+        // Instance is optional. Parse it if it's there, but leave the closing ] intact
+        let mut instance = {
+            let maybe_instance = if bytes.peek() == Some(&b':') {
+                bytes.next().expect("Byte was peeked");
+
+                let mut acc = u64::from(digit_from_ascii(bytes.next()?)?);
+                while let Some(digit) = bytes.peek().copied().and_then(digit_from_ascii) {
+                    bytes.next().expect("Byte was peeked");
+                    acc = acc.checked_mul(10)?;
+                    acc = acc.checked_add(u64::from(digit))?;
+                }
+
+                Some(Instance::from_u64(acc).unwrap_or(Instance::Invalid))
+            } else {
+                None
+            };
+
+            match (maybe_instance, type_char) {
+                (None, 'U') => Instance::Desktop,
+                (None, _) | (_, 'T' | 'g') => Instance::All,
+                (Some(instance), _) => instance,
+            }
+        };
+
         if let Some(i) = flag {
-            instance = Some(i);
+            instance = i;
         }
 
-        Ok(Self::new(
-            account_id,
-            instance.ok_or_else(Self::err)?,
-            account_type,
-            universe,
-        ))
+        if bytes.next() != Some(b']') {
+            return None;
+        }
+
+        Some(Self::new(account_id, instance, account_type, universe))
```

Comment:
This is ok as is, but should be able to implement this function without peekable just switching on the first non-digit character if you'd like to try.

Reply:
I think I would prefer to keep the peekable iterator for the steam3 parsing if you're fine with that. It keeps the parsing simpler, and is still faster than parsing with the regex

Reply:
Sure